File: c/pdfa_extension/checkconformance/checkconformance.c
/* checkconformance -- C port of
 * examples\Vb6\pdfa_extension\checkconformance\checkconformance.bas
 * Imports a PDF, converts it to PDF/A-3b via CheckConformance with
 * font-not-found and ICC-replacement callbacks, then writes the result. */
#include <stdio.h>
#include <string.h>
#include "lumaspdf.h"
#include "../../_common.h"

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

/* Data is the PDF handle passed as UserData to CheckConformance. */
static SI32 PDF_CALL FontNotFoundProc(void* Data, void* PDFFont, const char* FontName,
                                      TFStyle Style, SI32 StdFontIndex, LBOOL IsSymbolFont)
{
    (void)FontName; (void)StdFontIndex; (void)IsSymbolFont;
    /* WeightFromStyle is not exposed by the flat wrapper -> replace with Arial preserving Style. */
    return pdfReplaceFontA((PPDF)Data, PDFFont, "Arial", (SI32)Style, 1);
}

static SI32 PDF_CALL ReplaceICCProfileProc(void* Data, TICCProfileType ProfileType, SI32 ColorSpace)
{
    switch (ProfileType) {
        case ictRGB:  return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, "../../../test_files/sRGB.icc");
        case ictCMYK: return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, "../../../test_files/ISOcoated_v2_bas.ICC");
        default:      return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, "../../../test_files/gray.icc");
    }
}

static int ConvertFile(PPDF pdf, SI32 ConvType, const char* InFile, const char* OutFile)
{
    SI32 i, n, retval;
    UI32 convFlags;
    TPDFError e;

    pdfCreateNewPDFA(pdf, "");
    pdfSetDocInfoA(pdf, diProducer, "");

    switch (ConvType) {
        case ctNormalize:      convFlags = coAllowDeviceSpaces; break;
        case ctPDFA_1b_2005:   convFlags = coDefault | coFlattenLayers; break;
        case ctPDFA_2b:
        case ctPDFA_2u:        convFlags = coDefault | coDeletePresentation; break;
        default:               convFlags = (coDefault | coDeletePresentation) & ~coDeleteEmbeddedFiles; break;
    }
    convFlags |= coCheckImages;
    convFlags |= coRepairDamagedImages;

    if (ConvType != ctNormalize) {
        pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage | ifPrepareForPDFA);
        pdfSetImportFlags2(pdf, if2UseProxy | if2DuplicateCheck);
    } else {
        pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage);
    }

    retval = pdfCheckConformance(pdf, ConvType, convFlags, pdf, FontNotFoundProc, ReplaceICCProfileProc);
    switch (retval) {
        case 1: pdfAddOutputIntentA(pdf, "../../../test_files/sRGB.icc"); break;
        case 2: pdfAddOutputIntentA(pdf, "../../../test_files/ISOcoated_v2_bas.ICC"); break;
        case 3: pdfAddOutputIntentA(pdf, "../../../test_files/gray.icc"); break;
    }

    memset(&e, 0, sizeof(e));
    e.StructSize = sizeof(e);
    n = pdfGetErrLogMessageCount(pdf);
    for (i = 0; i < n; i++) {
        pdfGetErrLogMessage(pdf, i, &e);
        if (e.Msg) printf("%s\n", e.Msg);
    }

    if (pdfHaveOpenDoc(pdf)) {
        if (pdfOpenOutputFileA(pdf, OutFile) == 0) return 0;
        return pdfCloseFile(pdf) != 0;
    }
    return 0;
}

int main(int argc, char** argv)
{
    PPDF pdf;
    char dir[1024], outFile[1100], cmap[1100];
    (void)argc;

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, PDFError);
    exedir(argv[0], dir, sizeof(dir));
    _snprintf(cmap, sizeof(cmap), "%s\\..\\..\\..\\Resource\\CMap", dir);
    pdfSetCMapDirA(pdf, cmap, lcmDelayed | lcmRecursive);
    _snprintf(outFile, sizeof(outFile), "%s\\out.pdf", dir);

    if (ConvertFile(pdf, ctPDFA_3b, "../../../../sample_multipage.pdf", outFile))
        printf("PDF file \"%s\" successfully created!\n", outFile);

    pdfDeletePDF(pdf);
    return 0;
}

File: csharp/pdfa_extension/checkconformance/checkconformance.cs
// checkconformance -- C# port of examples\Vb6\pdfa_extension\checkconformance\checkconformance.bas
// Imports a PDF and converts it to PDF/A-3b via CheckConformance, using callbacks
// to replace missing fonts and ICC profiles, then writes the result.
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class CheckConformance
{
    static readonly string TestFiles = System.IO.Path.Combine(System.AppContext.BaseDirectory, "..", "..", "..", "..", "test_files");

    static TErrorProc _err = PDFError;
    static TOnFontNotFoundProc _fnf = FontNotFoundProc;
    static TOnReplaceICCProfile _icc = ReplaceICCProfileProc;

    static int PDFError(IntPtr data, int errCode, string errMessage, int errType)
    {
        Console.WriteLine(errMessage);
        return 0;
    }

    // Data is the PDF handle passed as UserData to CheckConformance.
    static int FontNotFoundProc(IntPtr data, IntPtr pdfFont, string fontName, int style, int stdFontIndex, bool isSymbolFont)
    {
        // Replace with Arial preserving Style.
        return LumasPdf.pdfReplaceFontW(data, pdfFont, "Arial", style, true);
    }

    static int ReplaceICCProfileProc(IntPtr data, TICCProfileType profileType, int colorSpace)
    {
        switch (profileType)
        {
            case TICCProfileType.ictRGB:  return LumasPdf.pdfReplaceICCProfileW(data, (uint)colorSpace, Path.Combine(TestFiles, "sRGB.icc"));
            case TICCProfileType.ictCMYK: return LumasPdf.pdfReplaceICCProfileW(data, (uint)colorSpace, Path.Combine(TestFiles, "ISOcoated_v2_bas.ICC"));
            default:                      return LumasPdf.pdfReplaceICCProfileW(data, (uint)colorSpace, Path.Combine(TestFiles, "gray.icc"));
        }
    }

    static bool ConvertFile(IntPtr pdf, TConformanceType convType, string inFile, string outFile)
    {
        LumasPdf.pdfCreateNewPDFW(pdf, "");             // The output file is opened later
        LumasPdf.pdfSetDocInfoW(pdf, TDocumentInfo.diProducer, "");

        uint convFlags;
        switch (convType)
        {
            case TConformanceType.ctNormalize:
                convFlags = LumasPdfConsts.coAllowDeviceSpaces;
                break;
            case TConformanceType.ctPDFA_1b_2005:
                convFlags = LumasPdfConsts.coDefault | LumasPdfConsts.coFlattenLayers;
                break;
            case TConformanceType.ctPDFA_2b:
            case TConformanceType.ctPDFA_2u:
                convFlags = LumasPdfConsts.coDefault | LumasPdfConsts.coDeletePresentation;
                break;
            default:
                // ctPDFA_3b, ctPDFA_4x, ZUGFeRD/FacturX: embedded files are allowed.
                convFlags = (LumasPdfConsts.coDefault | LumasPdfConsts.coDeletePresentation) & ~LumasPdfConsts.coDeleteEmbeddedFiles;
                break;
        }

        convFlags |= LumasPdfConsts.coCheckImages;
        convFlags |= LumasPdfConsts.coRepairDamagedImages;

        if (convType != TConformanceType.ctNormalize)
        {
            LumasPdf.pdfSetImportFlags(pdf, LumasPdfConsts.ifImportAll | LumasPdfConsts.ifImportAsPage | LumasPdfConsts.ifPrepareForPDFA);
            LumasPdf.pdfSetImportFlags2(pdf, LumasPdfConsts.if2UseProxy | LumasPdfConsts.if2DuplicateCheck);
        }
        else
        {
            LumasPdf.pdfSetImportFlags(pdf, LumasPdfConsts.ifImportAll | LumasPdfConsts.ifImportAsPage);
            LumasPdf.pdfSetImportFlags2(pdf, LumasPdfConsts.if2UseProxy | LumasPdfConsts.if2DuplicateCheck | LumasPdfConsts.if2Normalize);
        }

        int retval = LumasPdf.pdfOpenImportFileW(pdf, inFile, (int)LumasPdfConsts.ptOpen, "");
        if (retval < 0)
        {
            Console.WriteLine("Could not open the import file (it may be encrypted)!");
            LumasPdf.pdfFreePDF(pdf);
            return false;
        }
        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0);
        LumasPdf.pdfCloseImportFile(pdf);

        retval = LumasPdf.pdfCheckConformance(pdf, (int)convType, convFlags, pdf, _fnf, _icc);
        switch (retval)
        {
            case 1: LumasPdf.pdfAddOutputIntentW(pdf, Path.Combine(TestFiles, "sRGB.icc")); break;
            case 2: LumasPdf.pdfAddOutputIntentW(pdf, Path.Combine(TestFiles, "ISOcoated_v2_bas.ICC")); break;
            case 3: LumasPdf.pdfAddOutputIntentW(pdf, Path.Combine(TestFiles, "gray.icc")); break;
        }

        TPDFError e = new TPDFError();
        e.StructSize = (uint)Marshal.SizeOf(typeof(TPDFError));
        int count = LumasPdf.pdfGetErrLogMessageCount(pdf);
        for (int i = 0; i < count; i++)
        {
            LumasPdf.pdfGetErrLogMessage(pdf, (uint)i, ref e);
            Console.WriteLine(Marshal.PtrToStringAnsi(e.Msg));
        }

        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
                return false;
            return LumasPdf.pdfCloseFile(pdf);
        }
        return false;
    }

    static void Main()
    {
        string dir = AppDomain.CurrentDomain.BaseDirectory;
        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        LumasPdf.pdfSetCMapDirW(pdf, Path.Combine(dir, "CMap"), LumasPdfConsts.lcmDelayed | LumasPdfConsts.lcmRecursive);

        string outFile = Path.Combine(dir, "out.pdf");
        string inFile = Path.Combine(dir, "dynapdf_help.pdf");
        if (ConvertFile(pdf, TConformanceType.ctPDFA_3b, inFile, outFile))
            Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");

        LumasPdf.pdfDeletePDF(pdf);
    }
}
